### ISSUE TITLE
question

### ISSUE THREAD

**Issue by Arnold1** [actgardner/gogen-avro#133]
Hi,

I have the following schema:
```
{
    "type": "record",
    "name": "DemoSchema",
    "fields": [
        {
            "name": "test_id",
            "type": [
                "null",
                "long"
            ],
            "default": null
        }
    ]
}
```

how can I create the go struct -> serialize it -> deserialize it and read back the go struct?

when I try the following it doesn't work, I'm not sure how to read the test_id from the deserialized data...could you please show?
```
// Create a new DemoSchema struct
test_id := avro.NewUnionNullLong()
test_id.SetLong(1)

// Create a new DemoSchema struct
demoStruct := avro.DemoSchema{
    Test_id:    testID,
}

// Serialize the struct to a byte buffer
var buf bytes.Buffer
fmt.Printf("Serializing struct: %#v\n", demoStruct)
demoStruct.Serialize(&buf)

// Deserialize the byte buffer back into a struct
newDemoStruct, err := avro.DeserializeDemoSchema(&buf)
if err != nil {
    fmt.Printf("Error deserializing struct: %v\n", err)
    return
}
fmt.Printf("Deserialized struct: %#v\n", newDemoStruct)

res := newDemoStruct.Test_id
fmt.Println("res:", res.Long)
```

cc @actgardner @dmlambea

**Comment by actgardner** [comment]
`SetLong` on a union type is a method for the VM to populate the _type_ that's selected in the union, not to set the Long value itself. If you print the value of `test_id` before you serialize, it's clear:

```
Serializing struct: &main.UnionNullLong{Null:(*types.NullVal)(nil), Long:0, UnionType:1}
```
`SetLong` sets the `UnionType` field, but the `Long` field is still 0. What you actually want to do is ignore all the `Set` methods altogether (they exist to fulfill an interface used by gogen-avro), and just set the fields on the struct:

```
// Set the UnionType to Long (as opposed to UnionNullLongTypeEnumNull)
test_id.UnionType = UnionNullLongTypeEnumLong

// Set the Long value in the union to 1
test_id.Long = 1
```

**Comment by Arnold1** [comment]
@actgardner thanks so much for your reply.

can you just verify the following sample, any error checks a could add?
```
// Create a new DemoSchema struct
test_id := avro.NewUnionNullLong()
test_id.UnionType = avro.UnionNullLongTypeEnumLong
test_id.Long = 1

fmt.Println("test_id_input:", test_id.Long)

// Create a new DemoSchema struct
demoStruct := avro.DemoSchema{
    Test_id:    testID,
}

// Serialize the struct to a byte buffer
var buf bytes.Buffer
fmt.Printf("Serializing struct: %#v\n", demoStruct)
demoStruct.Serialize(&buf)

// Deserialize the byte buffer back into a struct
newDemoStruct, err := avro.DeserializeDemoSchema(&buf)
if err != nil {
    fmt.Printf("Error deserializing struct: %v\n", err)
    return
}
fmt.Printf("Deserialized struct: %#v\n", newDemoStruct)

test_id_union := newDemoStruct.Test_id // Question: can I check for nil here? or can I directly read newDemoStruct.Test_id.Long ?
fmt.Println("test_id_output:", test_id_union.Long)
```

**Comment by actgardner** [comment]
> Question: can I check for nil here? or can I directly read newDemoStruct.Test_id.Long ?

You should check the value of `demoStruct.test_id.UnionType` - if it's `UnionNullLongTypeEnumNull`, the enum value is `null`. If it's `UnionNullLongTypeEnumLong` then you can just read `demoStruct.test_id.Long`. In the current generated code (v6) `demoStruct.test_id` shouldn't ever be null after deserialization unless there's an error.

> lets say I have avro.DemoSchema with 100 variables, will I need to initialize them all before I call demoStruct.Serialize? is there a way to initialize them automatically somehow?

Right now you need to set values for the all the fields, they won't be initialized to the Avro schema defaults. I *think* it would make sense for the constructor method to set all the fields to the Avro defaults, but it doesn't currently. I might experiment with this. As it stands defaults exist mostly as a way to deserialize an older record into a newer schema with added fields.

> what is this example good for - can't I simply just store the bytes.Buffer into a file? 

I mean, you can. The serialized bytes of an Avro record are meaningless without a schema because it's an untagged format. The best blog post about binary data serialization formats is [this one by Martin Kleppman](https://martin.kleppmann.com/2012/12/05/schema-evolution-in-avro-protocol-buffers-thrift.html). I would recommend reading the Avro section at least.

In short, OCF provides a standardized way to:

- know that a file actually contains Avro data instead of just line noise
- store the schema together with a set of records
- compress Avro datasets
- work with datasets that are too big to fit into memory

There's also Single-Object Encoding which is designed to work with a schema registry for transmitting individual records over the network and through systems like Kafka.

> what can I do about this error?

I don't know, the golang mailing list might be able to help? My understanding is vendoring isn't recommended with Go modules.

**Comment by Arnold1** [comment]
@actgardner thank you for your reply.

> Right now you need to set values for the all the fields, they won't be initialized to the Avro schema defaults. I think it would make sense for the constructor method to set all the fields to the Avro defaults, but it doesn't currently. I might experiment with this. As it stands defaults exist mostly as a way to deserialize an older record into a newer schema with added fields.

can't I do the following? I call demoStruct.SetDefault and for the variable I want to initialize to e.g. Long I do the following?
```
demoStruct := avro.NewDemoSchema()
     for i := 0; i <= 100; i++ {
        demoStruct.SetDefault(i)
     }
     demoStruct.test_id.UnionType = avro.UnionNullLongTypeEnumLong
     demoStruct.test_id.Long = 1
```

**Comment by Arnold1** [comment]
> > Question: can I check for nil here? or can I directly read newDemoStruct.Test_id.Long ?
> 
> You should check the value of `demoStruct.test_id.UnionType` - if it's `UnionNullLongTypeEnumNull`, the enum value is `null`. If it's `UnionNullLongTypeEnumLong` then you can just read `demoStruct.test_id.Long`. In the current generated code (v6) `demoStruct.test_id` shouldn't ever be null after deserialization unless there's an error.
> 
> > lets say I have avro.DemoSchema with 100 variables, will I need to initialize them all before I call demoStruct.Serialize? is there a way to initialize them automatically somehow?
> 
> Right now you need to set values for the all the fields, they won't be initialized to the Avro schema defaults. I _think_ it would make sense for the constructor method to set all the fields to the Avro defaults, but it doesn't currently. I might experiment with this. As it stands defaults exist mostly as a way to deserialize an older record into a newer schema with added fields.
```
can't I do the following? I call demoStruct.SetDefault and for the variable I want to initialize to e.g. Long I do the following?

demoStruct := avro.NewDemoSchema()
     for i := 0; i <= 100; i++ {
        demoStruct.SetDefault(i)
     }
     demoStruct.test_id.UnionType = avro.UnionNullLongTypeEnumLong
     demoStruct.test_id.Long = 1
```
> 
> > what is this example good for - can't I simply just store the bytes.Buffer into a file?
> 
> I mean, you can. The serialized bytes of an Avro record are meaningless without a schema because it's an untagged format. The best blog post about binary data serialization formats is [this one by Martin Kleppman](https://martin.kleppmann.com/2012/12/05/schema-evolution-in-avro-protocol-buffers-thrift.html). I would recommend reading the Avro section at least.
> 
> In short, OCF provides a standardized way to:
> 
> * know that a file actually contains Avro data instead of just line noise
> * store the schema together with a set of records
> * compress Avro datasets
> * work with datasets that are too big to fit into memory
> 
> There's also Single-Object Encoding which is designed to work with a schema registry for transmitting individual records over the network and through systems like Kafka.
> 
> > what can I do about this error?
> 
> I don't know, the golang mailing list might be able to help? My understanding is vendoring isn't recommended with Go modules.

**Comment by actgardner** [comment]
> what happens if demoStruct.test_id.UnionType is UnionNullLongTypeEnumNull and I read demoStruct.test_id.Long?

It'll always be 0 because that's the Golang default value when the struct is initialized.

> can't I do the following? I call demoStruct.SetDefault for all the 100 variables - and for the variable I want to initialize myself e.g. test_id I do the following?

Yes, you could manually call `SetDefault` on every field if you only want to set a subset of fields.

**Comment by Arnold1** [comment]
thank u

**Comment by actgardner** [comment]
There's a lot of improvements in v7 that I've been working on for the past week. Otherwise the Issues section is a good place to see what features are being requested. Support for the confluent schema registry and logical types might be in the future. PRs are always welcome.

I'm going to close this issue out since we've solved your original problem.
